// tide/component.js
'use strict';

exports.__esModule = true;
exports.__RewireAPI__ = exports.__ResetDependency__ = exports.__set__ = exports.__Rewire__ = exports.__GetDependency__ = exports.__get__ = undefined;

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _shallowequal = require('shallowequal');

var _shallowequal2 = _interopRequireDefault(_shallowequal);

var _lodash = require('lodash.mapvalues');

var _lodash2 = _interopRequireDefault(_lodash);

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var NOT_KEY_PATH_PROPS = ['children', 'tide', 'tideOptions', 'key', 'ref'];

function omit(object, func) {
  return Object.keys(object).reduce(function (obj, key) {
    var _extends2;

    var value = object[key];
    return func(key, value) ? obj : _extends({}, obj, (_extends2 = {}, _extends2[key] = value, _extends2));
  }, {});
}

var excludedProps = _get__('NOT_KEY_PATH_PROPS').reduce(function (val, prop) {
  val[prop] = true;
  return val;
}, {});

var TideComponent = function (_get__$Component) {
  _inherits(TideComponent, _get__$Component);

  function TideComponent(props) {
    _classCallCheck(this, TideComponent);

    var _this = _possibleConstructorReturn(this, _get__$Component.apply(this, arguments));

    var tide = _this.getTide();
    var keyPaths = _this.getKeyPaths(props, tide);
    _this._componentTide = _extends({
      keyPaths: keyPaths,
      options: _this.getTideOptions()
    }, tide.getComponentProps());

    _this.state = _this.getPropsFromKeyPaths(keyPaths, tide);
    return _this;
  }

  TideComponent.prototype.getChildContext = function getChildContext() {
    return { tide: this.getTide(), tideOptions: this.getTideOptions() };
  };

  TideComponent.prototype.componentWillMount = function componentWillMount() {
    this._onStateChange = this.onStateChange.bind(this);
    this.getTide().onChange(this._onStateChange);
  };

  TideComponent.prototype.componentWillUnmount = function componentWillUnmount() {
    this._isUnmounting = true;
    this.getTide().offChange(this._onStateChange);
  };

  TideComponent.prototype.onStateChange = function onStateChange() {
    if (this._isUnmounting) return;
    var tide = this.getTide();
    var newState = this.getPropsFromKeyPaths(this.getKeyPaths(this.props, tide), tide);
    if (this.hasStaleState(newState)) {
      this.setState(newState);
    }
  };

  TideComponent.prototype.getKeyPaths = function getKeyPaths(props, tide) {
    var _this2 = this;

    var keyPaths = _get__('omit')(props, function (key) {
      return _get__('excludedProps')[key];
    });
    keyPaths = _get__('mapValues')(keyPaths, function (val, key) {
      var value = typeof val === 'function' ? val(tide.getState(), _this2.getTideOptions()) : val;
      if (Array.isArray(value)) return value;
      if (value === true) return [key];
      return value.split('.');
    });
    return keyPaths;
  };

  TideComponent.prototype.getPropsFromKeyPaths = function getPropsFromKeyPaths(keyPaths, tide) {
    var state = tide.getState();
    return _get__('mapValues')(keyPaths, function (value) {
      return state.getIn(value);
    });
  };

  TideComponent.prototype.getTide = function getTide() {
    return this.props.tide instanceof _get__('Tide') ? this.props.tide : this.context.tide;
  };

  TideComponent.prototype.getTideOptions = function getTideOptions() {
    return this.props.tideOptions || this.context.tideOptions || {};
  };

  TideComponent.prototype.getChildProps = function getChildProps() {
    return _extends({}, _get__('omit')(this.state, function (_, value) {
      return value === undefined;
    }), { tide: this._componentTide });
  };

  TideComponent.prototype.hasStaleState = function hasStaleState(newState) {
    return !_get__('shallowEqual')(this.state, newState);
  };

  TideComponent.prototype.render = function render() {
    return this.props.children(this.getChildProps());
  };

  return TideComponent;
}(_get__('React').Component);

exports.default = TideComponent;


if (process.env.NODE_ENV !== 'production') {
  _get__('TideComponent').displayName = 'TideComponent';
  _get__('TideComponent').propTypes = {
    tide: _get__('PropTypes').object
  };
}

var contextTypes = { tide: _get__('PropTypes').object, tideOptions: _get__('PropTypes').object };
_get__('TideComponent').contextTypes = _get__('contextTypes');
_get__('TideComponent').childContextTypes = _get__('contextTypes');

var _RewiredData__ = Object.create(null);

var INTENTIONAL_UNDEFINED = '__INTENTIONAL_UNDEFINED__';
var _RewireAPI__ = {};

(function () {
  function addPropertyToAPIObject(name, value) {
    Object.defineProperty(_RewireAPI__, name, {
      value: value,
      enumerable: false,
      configurable: true
    });
  }

  addPropertyToAPIObject('__get__', _get__);
  addPropertyToAPIObject('__GetDependency__', _get__);
  addPropertyToAPIObject('__Rewire__', _set__);
  addPropertyToAPIObject('__set__', _set__);
  addPropertyToAPIObject('__reset__', _reset__);
  addPropertyToAPIObject('__ResetDependency__', _reset__);
  addPropertyToAPIObject('__with__', _with__);
})();

function _get__(variableName) {
  if (_RewiredData__ === undefined || _RewiredData__[variableName] === undefined) {
    return _get_original__(variableName);
  } else {
    var value = _RewiredData__[variableName];

    if (value === INTENTIONAL_UNDEFINED) {
      return undefined;
    } else {
      return value;
    }
  }
}

function _get_original__(variableName) {
  switch (variableName) {
    case 'NOT_KEY_PATH_PROPS':
      return NOT_KEY_PATH_PROPS;

    case 'omit':
      return omit;

    case 'excludedProps':
      return excludedProps;

    case 'mapValues':
      return _lodash2.default;

    case 'Tide':
      return _base2.default;

    case 'shallowEqual':
      return _shallowequal2.default;

    case 'React':
      return _react2.default;

    case 'TideComponent':
      return TideComponent;

    case 'PropTypes':
      return _propTypes2.default;

    case 'contextTypes':
      return contextTypes;
  }

  return undefined;
}

function _assign__(variableName, value) {
  if (_RewiredData__ === undefined || _RewiredData__[variableName] === undefined) {
    return _set_original__(variableName, value);
  } else {
    return _RewiredData__[variableName] = value;
  }
}

function _set_original__(variableName, _value) {
  switch (variableName) {}

  return undefined;
}

function _update_operation__(operation, variableName, prefix) {
  var oldValue = _get__(variableName);

  var newValue = operation === '++' ? oldValue + 1 : oldValue - 1;

  _assign__(variableName, newValue);

  return prefix ? newValue : oldValue;
}

function _set__(variableName, value) {
  if ((typeof variableName === 'undefined' ? 'undefined' : _typeof(variableName)) === 'object') {
    Object.keys(variableName).forEach(function (name) {
      _RewiredData__[name] = variableName[name];
    });
  } else {
    if (value === undefined) {
      _RewiredData__[variableName] = INTENTIONAL_UNDEFINED;
    } else {
      _RewiredData__[variableName] = value;
    }

    return function () {
      _reset__(variableName);
    };
  }
}

function _reset__(variableName) {
  delete _RewiredData__[variableName];
}

function _with__(object) {
  var rewiredVariableNames = Object.keys(object);
  var previousValues = {};

  function reset() {
    rewiredVariableNames.forEach(function (variableName) {
      _RewiredData__[variableName] = previousValues[variableName];
    });
  }

  return function (callback) {
    rewiredVariableNames.forEach(function (variableName) {
      previousValues[variableName] = _RewiredData__[variableName];
      _RewiredData__[variableName] = object[variableName];
    });
    var result = callback();

    if (!!result && typeof result.then == 'function') {
      result.then(reset).catch(reset);
    } else {
      reset();
    }

    return result;
  };
}

var _typeOfOriginalExport = typeof TideComponent === 'undefined' ? 'undefined' : _typeof(TideComponent);

function addNonEnumerableProperty(name, value) {
  Object.defineProperty(TideComponent, name, {
    value: value,
    enumerable: false,
    configurable: true
  });
}

if ((_typeOfOriginalExport === 'object' || _typeOfOriginalExport === 'function') && Object.isExtensible(TideComponent)) {
  addNonEnumerableProperty('__get__', _get__);
  addNonEnumerableProperty('__GetDependency__', _get__);
  addNonEnumerableProperty('__Rewire__', _set__);
  addNonEnumerableProperty('__set__', _set__);
  addNonEnumerableProperty('__reset__', _reset__);
  addNonEnumerableProperty('__ResetDependency__', _reset__);
  addNonEnumerableProperty('__with__', _with__);
  addNonEnumerableProperty('__RewireAPI__', _RewireAPI__);
}

exports.__get__ = _get__;
exports.__GetDependency__ = _get__;
exports.__Rewire__ = _set__;
exports.__set__ = _set__;
exports.__ResetDependency__ = _reset__;
exports.__RewireAPI__ = _RewireAPI__;
